### PR TITLE
Lambda Dynamo saga sample

### PR DESCRIPTION
**Pull request by kentdr**

### CODE REVIEW

File: samples/dynamodb/lambda/DynamoDB_1/Sales/BillCustomerHandler.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+{
+  public class BillCustomerHandler : IHandleMessages<OrderReceived>
+  {
+    ILog log = LogManager.GetLogger<StageInventoryHandler>();
+
+    public async Task Handle(OrderReceived message, IMessageHandlerContext context)
+    {
+      log.Info("Billing customer.");
+
+      await Task.Delay(TimeSpan.FromSeconds(Random.Shared.Next(1, 15)), CancellationToken.None);
```
Comment (danielmarbach) [inline samples/dynamodb/lambda/DynamoDB_1/Sales/BillCustomerHandler.cs:18]: 15 seems excessive for a sample

```suggestion
      await Task.Delay(TimeSpan.FromSeconds(Random.Shared.Next(1, 5)), CancellationToken.None);
```

File: samples/dynamodb/lambda/DynamoDB_1/Sales/OrderSaga.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+      var sendOptions = new SendOptions();
+      sendOptions.RouteToThisEndpoint();
+
+      await context.Publish(new OrderReceived()
+      {
+        OrderId = message.OrderId
+      });
+
+      // Delay 10 seconds to allow time for order cancellation
+      await RequestTimeout<CancelOrder>(context, DateTimeOffset.UtcNow.AddSeconds(15));
```
Comment (danielmarbach) [inline samples/dynamodb/lambda/DynamoDB_1/Sales/OrderSaga.cs:42]: or change the comment

```suggestion
      await RequestTimeout<CancelOrder>(context, DateTimeOffset.UtcNow.AddSeconds(10));
```

File: samples/dynamodb/lambda/DynamoDB_1/Sales/Sales.csproj [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
+    <!-- Generate ready to run images during publishing to improve cold start time. -->
+    <PublishReadyToRun>true</PublishReadyToRun>
+    <LangVersion>8.0</LangVersion>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Amazon.Lambda.Core" Version="2.1.0" />
+    <PackageReference Include="Amazon.Lambda.Serialization.SystemTextJson" Version="2.3.1" />
+    <PackageReference Include="NServiceBus.AwsLambda.SQS" Version="1.0.0" />
+    <PackageReference Include="NServiceBus.Persistence.DynamoDB" Version="1.0.0" />
```
Comment (danielmarbach) [inline samples/dynamodb/lambda/DynamoDB_1/Sales/Sales.csproj:17]:
```suggestion
    <PackageReference Include="Amazon.Lambda.Core" Version="2.*" />
    <PackageReference Include="Amazon.Lambda.Serialization.SystemTextJson" Version="2.*" />
    <PackageReference Include="NServiceBus.AwsLambda.SQS" Version="1.*" />
    <PackageReference Include="NServiceBus.Persistence.DynamoDB" Version="1.*" />
```

Comment (danielmarbach) [reply samples/dynamodb/lambda/DynamoDB_1/Sales/Sales.csproj:17]: I realize we have the SystemTextJson pinned to a minor. Maybe we can align that as well https://github.com/Particular/docs.particular.net/blob/master/samples/hosting/aws-lambda-sqs/SQSLambda_1/AwsLambda.SQSTrigger/AwsLambda.SQSTrigger.csproj#L10

File: samples/dynamodb/lambda/DynamoDB_1/Sales/StageInventoryHandler.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+{
+  public class StageInventoryHandler : IHandleMessages<OrderReceived>
+  {
+    ILog log = LogManager.GetLogger<StageInventoryHandler>();
+
+    public async Task Handle(OrderReceived message, IMessageHandlerContext context)
+    {
+      log.Info("Staging inventory.");
+
+      await Task.Delay(TimeSpan.FromSeconds(Random.Shared.Next(1, 15)), CancellationToken.None);
```
Comment (danielmarbach) [inline samples/dynamodb/lambda/DynamoDB_1/Sales/StageInventoryHandler.cs:19]:
```suggestion
      await Task.Delay(TimeSpan.FromSeconds(Random.Shared.Next(1, 5)), CancellationToken.None);
```

File: samples/dynamodb/lambda/DynamoDB_1/Sales/aws-lambda-tools-defaults.json [hunk 5]
```diff
@@ -0,0 +1,14 @@
+{
+  "Information": [
+    "This file provides default values for the deployment wizard inside Visual Studio and the AWS Lambda commands added to the .NET Core CLI.",
+    "To learn more about the Lambda commands with the .NET Core CLI execute the following command at the command line in the project root directory.",
+    "dotnet lambda help",
+    "All the command line options for the Lambda command can be specified in this file."
+  ],
+  "profile": "default",
+  "region": "",
+  "configuration": "Release",
```
Comment (danielmarbach) [inline samples/dynamodb/lambda/DynamoDB_1/Sales/aws-lambda-tools-defaults.json:10]:
```suggestion
  "configuration": "Debug",
```

File: samples/dynamodb/lambda/sample.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+snippet: EndpointSetup
+
+The same class defines the AWS Lambda, which hosts the NServiceBus endpoint. The `ProcessOrder` method hands off processing of the message to NServiceBus:
+
+snippet: FunctionHandler
+
+Meanwhile, the `OrderSaga`, hosted within the AWS Lambda project, is a regular NServiceBus Saga which is also capable of sending messages receiving messages itself.
+
+snippet: OrderSaga
+
```
Comment (danielmarbach) [inline samples/dynamodb/lambda/sample.md:63]: Would it make sense to include a view of the dynamodb data as a screenshot?

File: samples/dynamodb/lambda/DynamoDB_1/Sales/OrderSaga.cs [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+       .ToMessage<CustomerBilled>(s => s.OrderId)
+       .ToMessage<InventoryStaged>(s => s.OrderId)
+       .ToMessage<CancelOrder>(s => s.OrderId);
+    }
+
+    public async Task Handle(PlaceOrder message, IMessageHandlerContext context)
+    {
+      log.Info($"Placing order: {Data.OrderId}");
+
+      var sendOptions = new SendOptions();
```
Comment (timbussmann) [inline samples/dynamodb/lambda/DynamoDB_1/Sales/OrderSaga.cs:33]: where is this sendoption used?

File: samples/dynamodb/lambda/sample.md [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+The sample includes a [`CloudFormation`](https://aws.amazon.com/cloudformation/aws-cloudformation-templates/) template, which will deploy the Lambda and create the necessary queues to run the sample.
+
+The [`Amazon.Lambda.Tools` CLI](https://github.com/aws/aws-lambda-dotnet) can be used to deploy the template to an AWS account.
+
+1. Install the [`Amazon.Lambda.Tools CLI`](https://github.com/aws/aws-lambda-dotnet#amazonlambdatools)
+1. Make sure an S3 bucket is available in the AWS region of choice
+
+## Running the sample
+
+INFO: It is not possible at this stage to use the AWS .NET Mock Lambda Test Tool to run the sample locally.
```
Comment (timbussmann) [inline samples/dynamodb/lambda/sample.md:26]: I'd say we should either explain why or maybe just remove this note?

Comment (danielmarbach) [reply samples/dynamodb/lambda/sample.md:26]: I would guess it is a note we have elsewhere too don't we?

Comment (kentdr) [reply samples/dynamodb/lambda/sample.md:26]: Yes in the simple AWS Lambda sample.

Comment (timbussmann) [reply samples/dynamodb/lambda/sample.md:26]: then we should explain why?

File: samples/dynamodb/lambda/DynamoDB_1/Messages/OrderShipped.cs [hunk 9]
```diff
@@ -0,0 +1,6 @@
+﻿using NServiceBus;
+
+public class OrderShipped : IEvent
+{
+  public string? OrderId { get; set; }
```
Comment (timbussmann) [inline samples/dynamodb/lambda/DynamoDB_1/Messages/OrderShipped.cs:5]: thinking a bit more about this, should these properties (on all messages) really be marked nullable? From a consumer standpoint this doesn't make much sense. Does this cause issues when creating the messages?

File: samples/dynamodb/lambda/DynamoDB_1/Sales/BillCustomerHandler.cs [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+using NServiceBus;
+using NServiceBus.Logging;
+
+public class BillCustomerHandler : IHandleMessages<OrderReceived>
+{
+  ILog log = LogManager.GetLogger<BillCustomerHandler>();
+
+  public async Task Handle(OrderReceived message, IMessageHandlerContext context)
+  {
+    log.Info("Billing customer.");
```
Comment (timbussmann) [inline samples/dynamodb/lambda/DynamoDB_1/Sales/BillCustomerHandler.cs:14]:
```suggestion
    log.Info($"Billing customer for order {message.OrderId}.");
```

File: samples/dynamodb/lambda/DynamoDB_1/Sales/OrderProcessor.cs [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+public class OrderProcessor
+{
+  #region EndpointSetup
+
+  static readonly AwsLambdaSQSEndpoint endpoint = new AwsLambdaSQSEndpoint(context =>
+  {
+    var endpointConfiguration = new AwsLambdaSQSEndpointConfiguration("Samples.DynamoDB.Lambda.Sales");
+
+
```
Comment (timbussmann) [inline samples/dynamodb/lambda/DynamoDB_1/Sales/OrderProcessor.cs:17]:
```suggestion
```

File: samples/dynamodb/lambda/DynamoDB_1/Sales/OrderProcessor.cs [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+  });
+
+  #endregion
+
+  #region FunctionHandler
+
+  public async Task ProcessOrder(SQSEvent eventData, ILambdaContext context)
+  {
+    context.Logger.Log("ProcessOrder was called");
```
Comment (timbussmann) [inline samples/dynamodb/lambda/DynamoDB_1/Sales/OrderProcessor.cs:39]: is ProcessOrder the right name given this endpoint handles different message types?

File: samples/dynamodb/lambda/DynamoDB_1/Sales/StageInventoryHandler.cs [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+using NServiceBus;
+using NServiceBus.Logging;
+
+public class StageInventoryHandler : IHandleMessages<OrderReceived>
+{
+  ILog log = LogManager.GetLogger<StageInventoryHandler>();
+
+  public async Task Handle(OrderReceived message, IMessageHandlerContext context)
+  {
+    log.Info("Staging inventory.");
```
Comment (timbussmann) [inline samples/dynamodb/lambda/DynamoDB_1/Sales/StageInventoryHandler.cs:14]:
```suggestion
    log.Info($"Staging inventory for order {message.OrderId}.");
```

**Review by timbussmann — APPROVED** [review]
looks good to me overall, made a few minor comments that might be worth looking into

File: samples/dynamodb/lambda/DynamoDB_1/Sales/OrderProcessor.cs (hunk 12 above)
Comment (timbussmann) [reply samples/dynamodb/lambda/DynamoDB_1/Sales/OrderProcessor.cs:38]: thinking more about this, I think this is good enough to leave as is.

File: samples/dynamodb/lambda/DynamoDB_1/Sales/OrderSaga.cs [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+  {
+    log.Info($"Placing order: {Data.OrderId}");
+
+    await context.Publish(new OrderReceived()
+    {
+      OrderId = message.OrderId
+    });
+
+    // Delay 5 seconds to allow time for order cancellation
+    await RequestTimeout<CancelOrder>(context, DateTimeOffset.UtcNow.AddSeconds(5));
```
Comment (timbussmann) [inline samples/dynamodb/lambda/DynamoDB_1/Sales/OrderSaga.cs:36]: I think we should remove the whole cancellation path completely to simplify the sample. Currently it's quite tricky to get a good timings to ensure that there are both successful orders and cancellations happening at a reasonable distribution. But in the end, it doesn't really end up providing much value in the context of the sample so why not just remove it?

File: samples/dynamodb/lambda/DynamoDB_1/ClientUI/Program.cs [hunk 15]
```diff
@@ -0,0 +1,47 @@
+﻿using System.Threading.Tasks;
+using System;
+using NServiceBus;
+
+class Program
+{
+  public static async Task Main(string[] args)
+  {
+    var endpointConfiguration = new EndpointConfiguration("Samples.DynamoDB.Lambda.ClientUI");
+    endpointConfiguration.SendFailedMessagesTo("Samples-DynamoDB-Lambda-Error");
```
Comment (timbussmann) [inline samples/dynamodb/lambda/DynamoDB_1/ClientUI/Program.cs:10]: configured this endpoint to use the same error queue

Comment (timbussmann) [reply samples/dynamodb/lambda/DynamoDB_1/ClientUI/Program.cs:10]: also disabled the installers for the client and added the necessary resources to the cloud formation template instead.

File: samples/dynamodb/lambda/DynamoDB_1/Sales/OrderSaga.cs (hunk 14 above)
Comment (danielmarbach) [reply samples/dynamodb/lambda/DynamoDB_1/Sales/OrderSaga.cs:36]: I'm good with that. Simpler is better.

File: samples/dynamodb/lambda/sample.md (hunk 6 above)
Comment (timbussmann) [reply samples/dynamodb/lambda/sample.md:63]: what kind of screenshot were you thinking about? Maybe a view of a records like this: 
image
?

Comment (danielmarbach) [reply samples/dynamodb/lambda/sample.md:63]: For example.

**Pull request merged**